### ISSUE TITLE
Union types

### ISSUE THREAD

**Issue by blois** [dart-lang/language#1222]
Currently Dart does not support function overloads based on parameter type and this makes it verbose/awkward to clone/wrap Javascript APIs which accept multiple types.

For example, WebSocket.send() can accept a String, a Blob or an ArrayBuffer.

In Dart, this can either be exposed as:
WebSocket.sendString, sendBlob, sendArrayBuffer which is verbose
or:
WebSocket.send(Dynamic data), which is no longer a self-documenting API.

It would be great to have some variant of union types, along the lines of JSDoc:
WebSocket.send(String|Blob|ArrayBuffer data);

**Comment by lrhn** [comment]
Deferring this discussion to #383

**Comment by easazade** [comment]
> > So what are the workarounds? Just use `dynamic` and manually check for correct types at runtime? (...) Are there any good workarounds? (...)
> 
> The common approach in the Dart/Flutter community in 2020 (that I personally see often) is to either use some code-generation libraries to generate union types e.g. [freezed](https://pub.dev/packages/freezed) or use library like [sealed_unions](https://pub.dev/packages/sealed_unions) or [union](https://pub.dev/packages/union).
> 
> Example union class in `freezed`:
> 
> ```dart
> @freezed
> abstract class Union with _$Union {
>   const factory Union(int value) = Data;
>   const factory Union.loading() = Loading;
>   const factory Union.error([String message]) = ErrorDetails;
> }
> ```
> 
> This of course isn't so nice as native support for unions, but gives you a lot of freedom in domain modeling and provides type safety.

Though freezed is a great library and i use id for defining data classes. using freezed for creating unions every time I need one is a pain. I have to define it then have to run `build_runner` which takes more than a minute on large projects.  all this comapred to just typing something like `File|Url` which takes a second? 
I rather don't use any union solution if it is going to take that much time just defining it.

**Comment by robertpiosik** [comment]
@easazade It's worth noting that `watch` argument makes subsequent code generation painless so I won't discourage anyone from generating unions with freezed as they will serve well.

**Comment by jodinathan** [comment]
> * Interface injection/traits.
> * Convenient union-wrappers with automatic coercion.

Could this work with JS bindings?

**Comment by lelabo-m** [comment]
#83 seems to consider Union a very different case from Sum Types.
It could be seen in there last messages that they might be related but the later seems easier to implement.

Should this topic be reopened ? Sum types might see light far earlier than union, and this subject will be certainly be buried as fixed ...

**Comment by ash0080** [comment]
I can't believe this discussion has been going on since 2012, and 10 years later it still hasn't happened. 😱

**Comment by mateusfccp** [comment]
> I can't believe this discussion has been going on since 2012, and 10 years later it still hasn't happened. 😱 

The team has not infinite resources and has to work with priorities.

**Comment by ash0080** [comment]
> > I can't believe this discussion has been going on since 2012, and 10 years later it still hasn't happened. 😱
> 
> The team has not infinite resources and has to work with priorities.

I agree with you very much, but personally I think that if there is a feature that can have a wider impact, it should have a higher priority, and there is no doubt that union type falls into this category.
I've looked at a lot of third-party packages, and about 1/3 of them make me think, "If there was a union type, there would be no need for this, or no need to write so much boilerplate code".

This is my immature opinion

**Comment by adifyr** [comment]
This is extremely high priority for us. I want this in Dart 2.18.

**Comment by ollydixon** [comment]
Here's a but from the main flutter socket library that was fixed because they had to use dynamic instead of a union type (wouldn't be an issue with TypeScript, Rust, C#, etc)

<img width="554" alt="Screenshot 2022-07-14 at 22 15 53" src="https://user-images.githubusercontent.com/6940726/179075929-9b0e5b23-1ad5-4c12-b960-4cc68fbf3e8b.png">

I can link many more reasons.

**Comment by duck-dev-go** [comment]
+1

**Comment by adifyr** [comment]
> > I can't believe this discussion has been going on since 2012, and 10 years later it still hasn't happened. 😱
> 
> The team has not infinite resources and has to work with priorities.

This feature should be top priority if you ask me. The benefit of implementing this would far outweigh the effort it would take to implement. If you'd implemented Data Classes and Union Types on their own - they would have released much faster & improved the language significantly.

But you're hogging up all your resources with the gigantic, tedious, beast of burden that is Static Metaprogramming - that nobody asked for (all we asked for was Data Classes) - which is still a couple of years from release.

**Comment by jodinathan** [comment]
>> beast of burden that is Static Metaprogramming (when all we really asked for was Data Classes)

Speak for yourself.  
Static metaprogramming can open a magnitude of development options

**Comment by ollydixon** [comment]
The static meta-programming looks interesting but it's value is much lower than union types in my opinion. I just don't see any use for it apart from making the language overly complex. Dart was nice and simple. Where was the voted on and why?

Union types + native JSON support is the only thing missing in Dart for the teams I work with. (mainly flutter and some server GRPC server stuff)

I see "dynamic" and the use of (someDynamic is someClass) for APIs & classes that include union types (usually from TypeScript with TRPC) and it sucks.

The union example given in the static meta explanation looks messy 🫣

**Comment by jodinathan** [comment]
>> I see "dynamic" and the use of (someDynamic is someClass) for APIs & classes that include union types (usually from TypeScript with TRPC) and it sucks in Dart.

Even using `dynamic` in Dart is a better Union Type usage than in TS.  
Dart has smart casting, thus you can `obj is SomeClass`, in TS you need to use the type guards and other messy stuff.  

In our projects, when we think "this could use a union type argument" we agree to rethink the entire logic of that component because it usually means something is wrong.  

I agree that Union Types is a must for Dart to prevail as all the kids on the block want it, however, static metaprogramming is insanely needed for big projects.

**Comment by ollydixon** [comment]
@jodinathan I'm dealing with projects (around 6 at the moment) some with +150k lines of code in Flutter (yes they are terrible). I still don't see a use-case for static meta-programming. Can you give me some examples of why we need it?

Unfortunately I see the need for Unions. Just a side note; Typescript you can do "is" as well without guards. Class meta can be attached to the JS for awhile now.

**Comment by jodinathan** [comment]
>> Can you give me some examples of why we need it?

 - Today is too much work to create and configure a builder for a simple source generation.  
 - With the augmentation feature we can expand classes without having to use `extends _$SomeStuff`.

We have internal server side and frontend side frameworks written in Dart that would benefit a lot from static metaprogramming.

We can automate several parts of the project that are currently manually coded.  
For example, a lot of our `enum` has a `translateName` method to be used on frontend:

```
enum BorderStyle {
  solid(borderSolid),
  dotted(borderDotted),
  dashed(borderDashed);

  const BorderStyle(this.translateName);

  static String borderSolid() => Intl.message('Solid',
      name: 'borderSolid', desc: 'Type "solid" of a border');

  static String borderDotted() => Intl.message('Dotted',
      name: 'borderDotted', desc: 'Type "dotted" of a border');

  static String borderDashed() => Intl.message('Dashed',
      name: 'borderDashed', desc: 'Type "dashed" of a border');

  final String Function() translateName;
}
```

A builder is too much work to automate that while a simple macro would definitely be an easy help.

Another example is that the translation package `Intl` can be entirely rewritten with it.

**Comment by ollydixon** [comment]
@jodinathan I really don't see this as a higher priority than a core language feature such as Unions though? Hmm.

Indeed using a builder can be annoying, we're using annotations with an RPS command to get around our JSON APIs in [TRPC](https://trpc.io/).

At least it's clear and not hiding anything from the developer. 
Dart is great for that reason, everything was simple and clear (easy to debug) 🚀🕺🏿

**Comment by jodinathan** [comment]
>> At least it's clear and not hiding anything from the developer.

Hence why Dart macro will use the new augmentation feature.  

From what I understood it will be like extending a class in a `part` file but without having to add the noisy `extends _$Foo`.

**Comment by dinbtechit** [comment]
static metaprogramming can be super helpful in so many cases. But I dont think it would be a great coding experience for unions though.

I dont think anything can beat the simplicity of this syntax.
```
typedef Padding = double | EdgeInsetsGeometry;
```

**Comment by ollydixon** [comment]
@dinbtechit exactly. 

@jodinathan the whole point of extends _$Foo is so that you can easily see it, tap on it, done.. don't hide things.
Extending classes gets REALLY messy, last decade I was a JAVA coder ...

What I see people wanting across teams (Flutter mainly which is basically what Dart is for):
1. Automatic JSON parsing into objects with readable errors. (Like TypeScript)
2. Unions, Partial, Omit types (and more, union being number 1)

I remember when we didn't have optionals (or non-nullable they call it).
They were debating if they should have optionals or not, insanity, a langauge without knowing if something is null or not...

Almost sounds like not knowing if something is union ...

Sometimes I feel they don't work on real world applications, how meta programing is above types is beyond me.

**Comment by mraleph** [comment]
> I dont think anything can beat the simplicity of this syntax.
> 
> ```
> typedef Padding = double | EdgeInsetsGeometry;
> ```

I think this is more of a counterexample which speaks against union types, rather then example which speaks in favor of them. Notice that this this type is 

1. not self-documenting i.e. what does `double` even mean here? Is it left, right, top, bottom, all of them, left and right, top and bottom, etc? I assume that it means padding from all sides, but it is not self-evident. 
2. there is an intersection between two parts of the union i.e. `EdgeInsets.all` can already encode padding from all sides.

In other words you can just use `EdgeInsetsGeometry` and you don't need such union type to begin with. Yep, the caller would need to write `EdgeInsets.all(10.0)` instead of just writing `10.0`. No, that's not a problem - verbosity here leads to clear and readable code.

**Comment by ollydixon** [comment]
@mraleph I think this is a bit off topic.

The issue isn't really local issues like this, he's just giving an example, but I think it's valid, I will give an answer below as to why it's obvious.

Some APIs force unions on us, and we are currently screwed; using dynamics with 'is' keyword and comments as a solution isn't great.

In dart we are missing Utility Types, [example](https://www.typescriptlang.org/docs/handbook/utility-types.html)
I'm not saying these are all good, but union definitely is. 
I can give examples where each one would be useful though.

---

But to the reply.

I've seen this before though using 'is' inside, example widget parameters.

```
SomeWidget(padding: 8) // Less verbose.
or
SomeWidget(padding: EdgeInsets.only(left: 8)
```

:-)

**Comment by jodinathan** [comment]
>> @jodinathan the whole point of extends _$Foo is so that you can easily see it, tap on it, done.. don't hide things.

@ollydixon hence why Dart team is creating the augmentation feature.  

From what I understand everything will be visible to the developer as it is now but better, for example when generating class source code we are not extending, but implementing it.  
If your generated class isn't used as mixin, you'll have to extend it and end up wasting your primary extension slot.

**Comment by dinbtechit** [comment]
@jodinathan - it definitely an interesting perspective. I think it's more applicable to the `num` datatype which is basically a combination of `int` & `double`.  Visibility can be helpful there because you're merging 2 different types.

However, the way how I look at unions are... you are just telling the compiler that in a given memory location you can either store TypeA or TypeB (i.e., `double` or `EdgeGeometry`). 

```dart
typedef Padding = double | EdgeGeometry;

// client code
Padding myPadding = 8;

// somewhere in the library when using the value

if (padding is double) {
  // Do something
} else if (padding is EdgeGeometry) {
  // Do something else
}
```
Notice - that we are not merging those 2 types. So when you are using the variable you still have to check the type. The advantage of this it will enforce the implementer to put in the right datatype at the development time or else they will get a compile error. I don't know if you need visibility into that because you already know what those types are.

For @mraleph earlier point:
> not self-documenting i.e. what does double even mean here?

That is true. But if you need to self-document you can do something like this.

```dart
typedef EdgeInsetsAllSides = double;
typedef Padding = EdgeInsetsAllSides | EdgeGeometry // Now it is verbose. 

// This is just an example I know there are other better ways you can do this. I am running into more complex scenarios for that I need this feature. I am not sure how to explain them here.
```

**Comment by jodinathan** [comment]
@dinbtechit I think you tagged the wrong user.  

But about your example:  

> ```dart
> typedef Padding = double | EdgeGeometry;
> 
> // client code
> Padding myPadding = 8;
> 
> // somewhere in the library when using the value
> 
> if (padding is double) {
>   // Do something
> } else if (padding is EdgeGeometry) {
>   // Do something else
> }
> ```
> 
> Notice - that we are not merging those 2 types. So when you are using the variable you still have to check the type. The advantage of this is it will enforce the implementer to put in the right datatype at the development time or else they will get a compile error. I don't know if you need visibility into that because you already know what those types are.
> 

What I dislike about union types is the feeling of bad performance that this gives me.  

In your example... Do we really need a runtime condition to something that should be able to be checked at compile time?  

ie:  

```dart
void geometryPadding(EdgeGeometry foo) {}

void padding(double foo) {}
```

Yeah it is a simple `if` but what if it is used too often by the renderer?  

Would be nice to use union types if the compiler can identify and tweak it at compile time:  

```dart
typedef Padding = double | EdgeGeometry;

void padd(Padding myPadding) {
  if (myPadding is double) {
    print('double! $myPadding');
  } else if (myPadding is EdgeGeometry) {
    print('geometry! $myPadding');
  }
}

dynamic foo = 10;

void main() {
  padd(8);
  padd(foo);
}

// compile the above into

void main() {
  print('double! 8');
  print('double! 10');
}
```

Sure the example above is simple, but if we keep all variable definition (that depends on the union types arguments) at the top of the method, the compiler should be able to aways identify and optimize the function calls.  
And if that is the case, then I think there would be almost no need to have method overloading

**Comment by paramsinghvc** [comment]
Why there's still no union type! It's quite surprising to me how this important feature is lacking in dart! 
I don't wanna use the freezed union thing for which I have to generate code. uhh

**Comment by searleser97** [comment]
I don't understand why this got closed. Is there a better dart alternative? What pattern should we follow instead ?

**Comment by EliahKagan** [comment]
@searleser97 This issue was closed as a duplicate of **#83**, which remains open.

(This this is so is not immediately obvious. But this issue was closed with https://github.com/dart-lang/language/issues/1222#issuecomment-691870412 which references #83, and the [`state-duplicate`](https://github.com/dart-lang/language/labels/state-duplicate) label was applied. See also https://github.com/dart-lang/language/issues/1222#issuecomment-396005754 for further context.)

**Comment by bvoq** [comment]
It would also allow us to move to effect-driven development, similarly to how https://effect.website/ works.
For more on effect types see:
https://koka-lang.github.io/koka/doc/index.html
https://zio.dev/
